keps/sig-node/20191127-cri-containerstats-enhancements.md

```diff
@@ -0,0 +1,212 @@
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+This enhancement is to extend `CpuUsage` and `MemoryUsage` fields in
+`ContainerStats`. And also add new `ProcessUsage`/`NetworkUsage` field in
+`ContainerStats`. The Container Runtime Engine will cover more metrics about
+managed container and let the metric collectors focus on node-level metrics or
+monitoring metrics.
+
+## Motivation
+
+Since more and more container runtimes shows up, like `kata-container`/
+`gVisor`/ `Firecracker`/ `Windows Containers`, the way to collect metrics from
+these containers is quite different from runC. For the third-party metric
+collectors, they have to maintain the knowledge about how to collect metrics
+from each kind of container runtimes. And more, kubelet codebase focus on
+core-metrics export (defined in [Kubernetes monitoring architecture#Terminology](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#terminology)),
+and there is no powerful standard interface for container level metrics.
+It will bring burden to end-users who wants to build powerful monitor pipeline
+(defined in [Kubernetes monitoring architecture#Executive Summary](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#executive-summary)) for different container runtimes.
+
+Based on this case, we can extend Container Runtime Interface(CRI) to define
+the container level metrics. Each container runtime engine, which implemented
+CRI, knows lifecycle and metrics about each managed container better than any
+external metric collector. For the third-party metric collectors, they just
+need to follow the `ContainerStats` API defined by CRI. With standard CRI,
+it is easy for the collector from monitor pipeline (defined in [Kubernetes monitoring architecture](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#executive-summary)) to collect metrics from different kind of runtimes.
+
+### Goals
+
+Enhance Container Runtime Interface(CRI) to provide more metrics about managed
+container to
+
+- Reduce resource usage of kubelet
+- Make eaiser integration for Monitoring pipeline
+
+## Proposal
+
+### User Stories
+
+As administrator, I provides sandbox runtime for applications in my cluster. I
+am not expertise of the sandbox runtime, but CRI implementation has created by
+domain expertise and CRI-API provides useful and powerful metrics for
+monitoring.
+
+To expose sandbox container's metrics, I only need to run Daemonset Monitoring
```

Comment:
This is too generic. Is this monitoring solution specific agent? Do every vendor needs to write their own agent now (I assume all vendors already have agents that would work with prometheus endpoint). How this agent  authorizes?

Reply:
Also if we _remove_ what the kubelet has then we at least need to have a replacement ready, not just say whatever the replacement would be "could be run if it existed" :)